ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java

```diff
@@ -0,0 +1,78 @@
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.function.udtf;
+
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.expression.tree.ColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.FunctionCall;
+import io.confluent.ksql.execution.function.UdtfUtil;
+import io.confluent.ksql.execution.plan.StreamFlatMap;
+import io.confluent.ksql.function.KsqlTableFunction;
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.schema.ksql.ColumnRef;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.OptionalInt;
+import org.apache.kafka.streams.kstream.ValueMapper;
+
+public class KudtfFlatMapper implements ValueMapper<GenericRow, Iterable<GenericRow>> {
+
+  private final StreamFlatMap streamFlatMap;
+
+  public KudtfFlatMapper(final StreamFlatMap streamFlatMap) {
+    this.streamFlatMap = streamFlatMap;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public Iterable<GenericRow> apply(final GenericRow row) {
+
+    /*
+    TODO this currently assumes function expression is a simple column ref expression
+     if we want to support more complex expressions we can use CodeGenRunner stuff
+     to evaluate expressions in a similar way to how SELECTs are evaluated.
+     */
+    final List<FunctionCall> functionCalls = streamFlatMap.getFunctionCalls();
+    final FunctionCall functionCall = functionCalls.get(0);
+    final ColumnReferenceExp exp = (ColumnReferenceExp) functionCall.getArguments().get(0);
+    final ColumnName columnName = exp.getReference().name();
+    final ColumnRef ref = ColumnRef.withoutSource(columnName);
+    final OptionalInt indexInInput = streamFlatMap.getInputSchema().valueColumnIndex(ref);
+    if (!indexInInput.isPresent()) {
+      throw new IllegalArgumentException("Can't find input column " + columnName);
+    }
+
+    // TODO we can cache all this (and above) so we don't look it up each time
+    final List<Object> unexplodedValue = row.getColumnValue(indexInInput.getAsInt());
+    final KsqlTableFunction tableFunction = UdtfUtil.resolveTableFunction(
+        streamFlatMap.getFunctionRegistry(),
+        functionCall,
+        streamFlatMap.getInputSchema()
+    );
+
+    final List<Object> list = tableFunction.flatMap(unexplodedValue);
+
+    final List<GenericRow> rows = new ArrayList<>();
+    for (Object val : list) {
+      final ArrayList<Object> arrayList = new ArrayList<>(row.getColumns());
```

Comment:
Consider:
 - nit: var type of `List` not `ArrayList`.
 - nit: name var after what it is, not what type it is
 - perf: pre-size array to known size

```suggestion
     final List<Object> outputRow = new ArrayList<>(row.getColumns().size() + 1);
      outputRow.addAll(row.getColumns());
```

Reply:
Again, I agree with most of these points, they are quite obvious and well known things.

General point: What I was looking for in this review is an overall validation, discussion and criticism of the overall architectural approach taken (and I can see there are some really good points you have made to that effect in the review, thank you), rather than a fine-toothed-comb forensic line by line analysis of the code. If you'd like to do that at some point, I don't mind, we can do Java 101 stuff but let's leave that to the end please, I think it would be more productive :)

Reply:
I've reviewed on the basis that the PR description and title didn't say anything to suggest you weren't intending to click the merge button once it was approved. 

I'm fine with providing a general review of approach, but it helps if you ask for it! ;)

Reply:
"completed"